File: wasm/bmxt-core/src/features/tabs_picker/group_target.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupTargetChoice {
    pub id: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveGroupTargetContext {
    pub pick_index: usize,
    pub choices: Vec<GroupTargetChoice>,
    pub new_group_sentinel: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedGroupTarget {
    pub create_new: bool,
    pub group_id: Option<i32>,
}

pub fn resolve_group_target(ctx: ResolveGroupTargetContext) -> Option<ResolvedGroupTarget> {
    let picked = ctx.choices.get(ctx.pick_index)?;
    if picked.id == ctx.new_group_sentinel {
        return Some(ResolvedGroupTarget {
            create_new: true,
            group_id: None,
        });
    }
    Some(ResolvedGroupTarget {
        create_new: false,
        group_id: Some(picked.id),
    })
}

File: lib/features/bmxt-core/tabs-picker/group-target.ts
export type GroupTargetChoice = { id: number }

export type ResolveGroupTargetContext = {
  pickIndex: number
  choices: GroupTargetChoice[]
  newGroupSentinel: number
}

export type ResolvedGroupTarget = {
  createNew: boolean
  groupId: number | null
}

export function resolveGroupTarget(ctx: ResolveGroupTargetContext): ResolvedGroupTarget | null {
  const picked = ctx.choices[ctx.pickIndex]
  if (!picked) return null
  if (picked.id === ctx.newGroupSentinel) {
    return { createNew: true, groupId: null }
  }
  return { createNew: false, groupId: picked.id }
}

export function resolveTabsPickerGroupTarget<TContext, TResult>(context: TContext): TResult {
  return resolveGroupTarget(context as ResolveGroupTargetContext) as TResult
}
